### PR TITLE
Reintroduce shared_paths

### PR DESCRIPTION
**Pull request by mattijn**
This PR reintroduce the function `pygeos.shared_paths()` as was discussed in https://github.com/pygeos/pygeos/issues/73.

The `shared_paths()` function returns a `pygeos.Geometry GEOMETRYCOLLECTION`. 

To access the shared paths that have the same direction in both inputs use index 0 in `pygeos.get_geometry()` using the result of `shared_paths()` and use index 1 to access the shared paths that have the opposite direction.

### CODE REVIEW

**Comment by caspervdw** [comment]
Thanks @mattijn for your contribution. I don't remember why I removed this function. It might have had some edge cases that lead to segfaults and I didn't have time to cover them.

Before this can be merged, this needs some work on unittests and documentation. Do you have time to have a look?

**Comment by mattijn** [comment]
I added a test and docstring including an example.

**Comment by jorisvandenbossche** [comment]
Thanks for the update!

Some other things:

- Could you add a test with None as input?
- The `axis` is not tested right now. But wondering: is this needed? Or how is it actually working? (the shared_path is working with 2 geoms, so in a reduction, is it doing that each time with the next geom and the result of the previous? Does that work with the 2-item GeometryCollection ?

**Comment by mattijn** [comment]
Added a test using `None`s and one using the `axis` parameter.

I was also surprised on the `axis` part, but it works with a `np.vstack` array-like input:
```python
arr1 = [g1, g2]
arr2 = [g1, g2]
geom = np.vstack((arr1, arr2)) # [[g1, g2], [g1, g2]]
pygeos.shared_paths(geom, axis=1)
```
```
array([<pygeos.Geometry GEOMETRYCOLLECTION (MULTILINESTRING ((0 0, 1 0)), MULTILINESTRING EMPTY)>,
       <pygeos.Geometry GEOMETRYCOLLECTION (MULTILINESTRING ((0 0, 1 0)), MULTILINESTRING EMPTY)>],
      dtype=object)
```
So no fancy stuff

**Comment by jorisvandenbossche** [comment]
Regarding None, it seems that if only the right operand is None, it doesn't give the expected behaviour (at least to me):

```
In [6]: g1 = pygeos.linestrings([(0, 0), (1, 0), (1, 1)]) 
   ...: g2 = pygeos.linestrings([(0, 0), (1, 0)])   

In [7]: pygeos.shared_paths(None, None) 

In [8]: pygeos.shared_paths(g1, None) 
Out[8]: <pygeos.Geometry LINESTRING (0 0, 1 0, 1 1)>

In [9]: pygeos.shared_paths(None, g1)
```

File: pygeos/test/test_linear.py [hunk 1]
```diff
@@ -64,3 +64,11 @@ def test_line_merge_geom_array():
     actual = pygeos.line_merge([line_string, multi_line_string])
     assert pygeos.equals(actual[0], line_string)
     assert pygeos.equals(actual[1], multi_line_string)
+
+def test_shared_paths_linestring():
+    g1 = pygeos.linestrings([(0, 0), (1, 0), (1, 1)])
+    g2 = pygeos.linestrings([(0, 0), (1, 0)])
+    actual1 = pygeos.shared_paths(g1, g2)
+    assert pygeos.equals(pygeos.get_geometry(actual1, 0), g2)
+    assert pygeos.shared_paths(None, None) == None
```
Comment (jorisvandenbossche) [inline pygeos/test/test_linear.py]:
```suggestion
    assert pygeos.shared_paths(None, None) is None
```

(in practice doesn't matter much, but for checking to None, using `is` is seen as better practice)

**Comment by jorisvandenbossche** [comment]
And for the axis behaviour: your example works as long as the axis you specify has only two elements. But once it has more, you get an error:

```
In [12]: pygeos.shared_paths([[g1, g2, g1], [g1, g2, g1]], axis=1) 
---------------------------------------------------------------------------
GEOSException                             Traceback (most recent call last)
<ipython-input-12-b4f3ff6754a6> in <module>
----> 1 pygeos.shared_paths([[g1, g2, g1], [g1, g2, g1]], axis=1)

~/scipy/repos/pygeos/pygeos/linear.py in shared_paths(geom1, geom2, axis)
    113     """    
    114     if geom2 is None:
--> 115         return lib.shared_paths.reduce(geom1, axis=axis)
    116     else:
    117         return lib.shared_paths(geom1, geom2)

GEOSException: IllegalArgumentException: Geometry is not lineal
```

This is because it first calculates the shared path of the first two items, resulting in a GeometryCollection, and then it calculates the shared path of this collection with the third item. But you can't pass GeometryCollection to this algo, it needs linear geometries.

So therefore (unless you have a good use case for it?), I would maybe suggest to leave out the `axis` behaviour for now. We can always easily add it back later if there is demand for it.

**Comment by mattijn** [comment]
Removed the `axis` parameter together with the `reduce` functionality.

**Comment by jorisvandenbossche** [comment]
@mattijn thanks for the update! 

I pushed some small stylistic edits (hope you are fine with that; that's a bit easier than giving a bunch of small nitpicky comments ;))

Looks good to me now. @caspervdw ?

**Comment by caspervdw** [comment]
Looks good to me too! Two questions:
- could you propagate `**kwargs` in the function?
- could you add a test that feeds a non-linestring to the function? It should raise.

**Comment by caspervdw** [comment]
@mattijn merging this, thanks again
